### PR TITLE
Added storage for direct filesystem references in code

### PR DESCRIPTION
**Pull request by ericvergnaud**
## Changes
On top of linting, collects DFSA records and stores them

### Linked issues
Progresses #2350 

### Functionality
- [x] added a new table `directfs_in_paths`

### Tests
- [x] added unit tests
- [x] updated integration tests
- [x] manually tested schema upgrade:

Screenshot 2024-09-13 at 16 00 19

### CODE REVIEW

**Comment by ericvergnaud** [comment]
Integration test fails due to lsql issue, tentative fix in https://github.com/databrickslabs/lsql/pull/266

**Comment by ericvergnaud** [comment]
@nfx we need to discuss 'unique' storage. Currently duplicates can be stored which is probably undesirable. Not sure what's the requirement ?

**Comment by ericvergnaud** [comment]
@nfx @asnare Following our call, here is my understanding of the requirement:
 - deduplication is not required since consumers can simply run a SELECT DISTINCT ...
 - data will be historicized in a dedicated table (separate PR)
 - DFSA record needs to be enriched with job id, task id, run id, run timestamp, source lineage and source timestamp
Not part of this PR:
Query DFSAs will be stored in a separate table (as part of https://github.com/databrickslabs/ucx/pull/2531)
DFSA Location -> Unity table mapping will need to be suggested, confirmed and stored (food for thought)

File: src/databricks/labs/ucx/source_code/graph.py [hunk 1]
```diff
@@ -289,12 +289,13 @@ class DependencyGraphContext:
     session_state: CurrentSessionState
 
 
-class Dependency(abc.ABC):
+class Dependency:
 
-    def __init__(self, loader: DependencyLoader, path: Path, inherits_context=True):
+    def __init__(self, loader: DependencyLoader, path: Path, inherits_context=True, lineage_str: str | None = None):
```
Comment (nfx) [inline src/databricks/labs/ucx/source_code/graph.py:294]:
```suggestion
    def __init__(self, loader: DependencyLoader, path: Path, inherits_context=True, lineage_str: list[Link] | None = None):
```

```
@dataclass
class Link:
    object_type: str
    object_id: str
```

let's use typed entries and not just strings. selialize to JSON on storage. If you want - do it in a separate PR.

File: src/databricks/labs/ucx/source_code/jobs.py [hunk 2]
```diff
@@ -441,3 +458,76 @@ def _process_dependency(
...
+        self, source: str, language: CellLanguage, path: Path, inherited_tree: Tree | None
+    ) -> Iterable[DirectFsAccess]:
+        notebook = Notebook.parse(path, source, language.language)
+        for cell in notebook.cells:
+            src_timestamp = int(path.stat().st_mtime)
+            src_id = str(path)
+            src_lineage = self.lineage_str
+            for dfsa in self._collect_from_source(cell.original_code, cell.language, path, inherited_tree):
+                yield dfsa.replace(
+                    source_type="NOTEBOOK", source_id=src_id, source_lineage=src_lineage, source_timestamp=src_timestamp
```
Comment (nfx) [inline src/databricks/labs/ucx/source_code/jobs.py]:
```suggestion
                    source_type="NOTEBOOK", source_id=src_id, source_lineage=src_lineage, source_timestamp=src_timestamp,
```
it reformats to multiple lines then

File: src/databricks/labs/ucx/source_code/jobs.py [hunk 3]
```diff
@@ -441,3 +458,76 @@ def _process_dependency(
...
+            iterable = self._collect_from_python(source, inherited_tree)
+        if iterable is None:
+            logger.warning(f"Language {language.name} not supported yet!")
+            return
+        src_id = str(path)
+        src_lineage = self.lineage_str
+        src_timestamp = int(path.stat().st_mtime)
+        for dfsa in iterable:
+            yield dfsa.replace(
+                source_type="FILE", source_id=src_id, source_lineage=src_lineage, source_timestamp=src_timestamp
```
Comment (nfx) [inline src/databricks/labs/ucx/source_code/jobs.py]:
```suggestion
                source_type="FILE", source_id=src_id, source_lineage=src_lineage, source_timestamp=src_timestamp,
```
reformat

File: src/databricks/labs/ucx/source_code/jobs.py [hunk 4]
```diff
@@ -441,3 +458,76 @@ def _process_dependency(
...
+            )
+
+    def _collect_from_python(self, source: str, inherited_tree: Tree | None) -> Iterable[DirectFsAccess]:
+        linter = DirectFsAccessPyLinter(self._session_state, prevent_spark_duplicates=False)
+        for dfsa_node in linter.collect_dfsas(source, inherited_tree):
+            yield dfsa_node.dfsa
+
+    def _collect_from_sql(self, source: str) -> Iterable[DirectFsAccess]:
```
Comment (nfx) [inline src/databricks/labs/ucx/source_code/jobs.py:536]: do we analyse the `spark.sql("SELECT * FROM csv.`/mnt/foo/bar`")`?

File: src/databricks/labs/ucx/source_code/linters/directfs.py [hunk 5]
```diff
@@ -141,26 +147,88 @@ def lint_tree(self, tree: Tree) -> Iterable[Advice]:
...
+        visitor.visit(tree.node)
+        yield from visitor.directfs_nodes
 
-class DirectFsAccessSqlLinter(SqlLinter):
 
-    def lint_expression(self, expression: Expression):
-        for table in expression.find_all(Table):
-            # Check table names for direct file system access
-            yield from self._check_dfsa(table)
+class DirectFsAccessSqlLinter(SqlLinter):
```
Comment (nfx) [inline src/databricks/labs/ucx/source_code/linters/directfs.py:156]: is it called from `DirectFsAccessPyLinter` for `spark.sql(XXX)` case? if not - it has to.

File: tests/unit/source_code/conftest.py [hunk 6]
```diff
@@ -55,3 +58,8 @@ def simple_dependency_resolver(mock_path_lookup: PathLookup) -> DependencyResolv
     notebook_resolver = NotebookResolver(NotebookLoader())
     import_resolver = ImportFileResolver(FileLoader(), allow_list)
     return DependencyResolver(library_resolver, notebook_resolver, import_resolver, import_resolver, mock_path_lookup)
+
+
+@pytest.fixture
+def mock_dfsa_crawlers() -> DirectFsAccessCrawlers:
```
Comment (nfx) [inline tests/unit/source_code/conftest.py:64]: use `create_autospec(DirectFsAccessCrawlers)` in every test that uses it

File: tests/unit/source_code/samples/functional/file-access/select_format.sql [hunk 7]
```diff
@@ -0,0 +1,3 @@
+-- Databricks notebook source
+-- ucx[direct-filesystem-access-in-sql-query:+0:0:+0:1024] The use of direct filesystem references is deprecated: hdfs://examples/src/main/resources/users.parquet
+SELECT * FROM parquet.`hdfs://examples/src/main/resources/users.parquet`
```
Comment (nfx) [inline tests/unit/source_code/samples/functional/file-access/select_format.sql:3]: `hdfs://` is not accessible from Databricks

**Review by nfx — CHANGES_REQUESTED** [review]

File: src/databricks/labs/ucx/source_code/graph.py [hunk 8]
```diff
@@ -624,3 +632,8 @@ def _log_walk_one(self, dependency: Dependency):
     def _process_dependency(
         self, dependency: Dependency, path_lookup: PathLookup, inherited_tree: Tree | None
     ) -> Iterable[T]: ...
+
+    @property
+    def lineage_str(self):
+        parts = [dependency.lineage_str for dependency in self._lineage]
+        return "->".join(parts)
```
[the remark this answers is not in the document]
Comment (ericvergnaud) [reply src/databricks/labs/ucx/source_code/graph.py:639]: Each lineage component is double-quoted

File: src/databricks/labs/ucx/source_code/jobs.py (hunk 4 above)
Comment (ericvergnaud) [reply src/databricks/labs/ucx/source_code/jobs.py:536]: yes that's done as part of pyspark linting

File: src/databricks/labs/ucx/source_code/graph.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (nfx) [reply src/databricks/labs/ucx/source_code/graph.py:639]: use JSON/concrete classes :)

File: tests/unit/source_code/conftest.py [hunk 9]
```diff
@@ -55,3 +58,8 @@ def simple_dependency_resolver(mock_path_lookup: PathLookup) -> DependencyResolv
     notebook_resolver = NotebookResolver(NotebookLoader())
     import_resolver = ImportFileResolver(FileLoader(), allow_list)
     return DependencyResolver(library_resolver, notebook_resolver, import_resolver, import_resolver, mock_path_lookup)
+
+
+@pytest.fixture
+def mock_dfsa_crawler() -> DirectFsAccessCrawler:
```
[the remark this answers is not in the document]
Comment (ericvergnaud) [reply tests/unit/source_code/conftest.py]: done

File: src/databricks/labs/ucx/source_code/directfs_access_crawler.py [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+        yield from self._backend.fetch(sql)
+
+    def _crawl(self) -> Iterable[Result]:
+        return []
+
+
+class DirectFsAccessCrawlers:
+
+    def __init__(self, backend: SqlBackend, schema: str):
+        self._backend = backend
```
[the remark this answers is not in the document]
Comment (ericvergnaud) [reply src/databricks/labs/ucx/source_code/directfs_access_crawler.py:42]: done

File: src/databricks/labs/ucx/source_code/graph.py (hunk 1 above)
Comment (ericvergnaud) [reply src/databricks/labs/ucx/source_code/graph.py:294]: Created https://github.com/databrickslabs/ucx/issues/2550

File: src/databricks/labs/ucx/source_code/jobs.py (hunk 2 above)
Comment (ericvergnaud) [reply src/databricks/labs/ucx/source_code/jobs.py]: this code is not the latest

File: src/databricks/labs/ucx/source_code/jobs.py (hunk 3 above)
Comment (ericvergnaud) [reply src/databricks/labs/ucx/source_code/jobs.py]: this code is not the latest

File: tests/unit/source_code/conftest.py (hunk 6 above)
Comment (ericvergnaud) [reply tests/unit/source_code/conftest.py:64]: done

File: tests/unit/source_code/samples/functional/file-access/select_format.sql (hunk 7 above)
Comment (ericvergnaud) [reply tests/unit/source_code/samples/functional/file-access/select_format.sql]: done

File: src/databricks/labs/ucx/source_code/graph.py (hunk 1 above)
Comment (nfx) [reply src/databricks/labs/ucx/source_code/graph.py:294]: can you address it in this PR or decouple lineage from it?

File: src/databricks/labs/ucx/source_code/linters/directfs.py (hunk 5 above)
Comment (ericvergnaud) [reply src/databricks/labs/ucx/source_code/linters/directfs.py:156]: No. We're not linting query in `spark.sql`. That seems to be an old bug and much broader bug. I suggest fixing that in a separate PR (with missing tests). See https://github.com/databrickslabs/ucx/issues/2551

File: tests/unit/source_code/conftest.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (ericvergnaud) [reply tests/unit/source_code/conftest.py]: but it raises `mock-no-usage` pylint error

File: src/databricks/labs/ucx/source_code/directfs_access_crawler.py [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+
+    def __init__(self, backend: SqlBackend, schema: str, table: str):
+        """
+        Initializes a DFSACrawler instance.
+
+        Args:
+            backend (SqlBackend): The SQL Execution Backend abstraction (either REST API or Spark)
+            schema: The schema name for the inventory persistence.
+        """
+        super().__init__(backend, "hive_metastore", schema, table, DirectFsAccess)
```
[the remark this answers is not in the document]
Comment (ericvergnaud) [reply src/databricks/labs/ucx/source_code/directfs_access_crawler.py]: done

File: src/databricks/labs/ucx/source_code/jobs.py [hunk 12]
```diff
@@ -412,20 +426,23 @@ def _lint_task(self, task: jobs.Task, job: jobs.Job, linted_paths: set[Path]) ->
             graph, linted_paths, self._path_lookup, task.task_key, session_state, self._migration_index
         )
         yield from walker
+        collector = DfsaCollectorWalker(graph, set(), self._path_lookup, session_state)
```
[the remark this answers is not in the document]
Comment (ericvergnaud) [reply src/databricks/labs/ucx/source_code/jobs.py:458]: Done

File: src/databricks/labs/ucx/source_code/graph.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (ericvergnaud) [reply src/databricks/labs/ucx/source_code/graph.py:639]: Done in PR #2556

File: src/databricks/labs/ucx/source_code/graph.py (hunk 1 above)
Comment (ericvergnaud) [reply src/databricks/labs/ucx/source_code/graph.py:294]: I can merge the PR but maybe you want to review it first ?

Comment (ericvergnaud) [reply src/databricks/labs/ucx/source_code/graph.py:294]: See https://github.com/databrickslabs/ucx/pull/2556

File: src/databricks/labs/ucx/source_code/linters/directfs.py (hunk 5 above)
Comment (ericvergnaud) [reply src/databricks/labs/ucx/source_code/linters/directfs.py:156]: See PR #2558

File: src/databricks/labs/ucx/source_code/jobs.py (hunk 4 above)
Comment (ericvergnaud) [reply src/databricks/labs/ucx/source_code/jobs.py:560]: Actually it's not. Fixed in #2558

**Comment by ericvergnaud** [comment]
blocked by https://github.com/databrickslabs/ucx/pull/2562

File: src/databricks/labs/ucx/source_code/base.py [hunk 13]
```diff
@@ -336,3 +337,77 @@ def is_a_notebook(path: Path, content: str | None = None) -> bool:
...
+    is_read: bool
+    is_write: bool
+    source_id: str = UNKNOWN
+    source_timestamp: int = -1
+    source_lineage: str = UNKNOWN
+    job_id: int = -1
+    job_name: str = UNKNOWN
+    task_key: str = UNKNOWN
+    assessment_start_timestamp: int = -1
+    assessment_end_timestamp: int = -1
```
Comment (nfx) [inline src/databricks/labs/ucx/source_code/base.py:358]:
```suggestion
    source_timestamp: datetime.datetime
    source_lineage: list[LineageAtom]
    job_id: int = -1
    job_name: str = UNKNOWN
    task_key: str = UNKNOWN
    assessment_start_timestamp: datetime.datetime
    assessment_end_timestamp: datetime.datetime
```

see https://github.com/databrickslabs/lsql/pull/272

I'm also struggling to find where `assessment_start_timestamp` and `assessment_end_timestamp` are used. we have the `source_timestamp` already, so why do we need two more timestamps?

File: src/databricks/labs/ucx/source_code/directfs_access_crawler.py [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+
+    def __init__(self, sql_backend: SqlBackend, schema: str):
+        self._sql_backend = sql_backend
+        self._schema = schema
+
+    def for_paths(self) -> _DirectFsAccessCrawler:
+        return _DirectFsAccessCrawler(self._sql_backend, self._schema, "direct_file_system_access_in_paths")
+
+    def for_queries(self) -> _DirectFsAccessCrawler:
+        return _DirectFsAccessCrawler(self._sql_backend, self._schema, "direct_file_system_access_in_queries")
```
Comment (nfx) [inline src/databricks/labs/ucx/source_code/directfs_access_crawler.py:49]: @asnare can you take a look at this code and align it with what you're working on?

File: src/databricks/labs/ucx/source_code/directfs_access_crawler.py [hunk 15]
```diff
@@ -0,0 +1,49 @@
...
+
+
+class DirectFsAccessCrawlers:
+
+    def __init__(self, sql_backend: SqlBackend, schema: str):
+        self._sql_backend = sql_backend
+        self._schema = schema
+
+    def for_paths(self) -> _DirectFsAccessCrawler:
+        return _DirectFsAccessCrawler(self._sql_backend, self._schema, "direct_file_system_access_in_paths")
```
Comment (nfx) [inline src/databricks/labs/ucx/source_code/directfs_access_crawler.py:46]: please add a view to union these two in src/databricks/labs/ucx/queries/views

File: src/databricks/labs/ucx/source_code/base.py (hunk 13 above)
Comment (ericvergnaud) [reply src/databricks/labs/ucx/source_code/base.py:358]: because there's never enough information to analyze issues in prod. Alternately we could have the run id of the workflow that generated these DFSAs but I couldn't find how to get that information from the code being executed by the workflow

File: src/databricks/labs/ucx/source_code/directfs_access_crawler.py (hunk 15 above)
Comment (ericvergnaud) [reply src/databricks/labs/ucx/source_code/directfs_access_crawler.py]: done

File: src/databricks/labs/ucx/source_code/directfs_access.py [hunk 16]
```diff
@@ -0,0 +1,136 @@
...
+    is_read: bool
+    is_write: bool
+    source_id: str = UNKNOWN
+    source_timestamp: datetime = datetime.fromtimestamp(0)
+    source_lineage: list[LineageAtom] = field(default_factory=list)
+    job_id: int = -1
+    job_name: str = UNKNOWN
+    task_key: str = UNKNOWN
+    assessment_start_timestamp: datetime = datetime.fromtimestamp(0)
+    assessment_end_timestamp: datetime = datetime.fromtimestamp(0)
```
[the remark this answers is not in the document]
Comment (ericvergnaud) [reply src/databricks/labs/ucx/source_code/directfs_access.py:40]: The issue is that they are populated gradually rather than at once.

File: src/databricks/labs/ucx/source_code/directfs_access.py [hunk 17]
```diff
@@ -0,0 +1,136 @@
...
+            self._append_records(dfsas)
+        except DatabricksError as e:
+            logger.error("Failed to store DFSAs", exc_info=e)
+
+    def _try_fetch(self) -> Iterable[DirectFsAccess]:
+        sql = f"SELECT * FROM {self.full_name}"
+        yield from self._backend.fetch(sql)
+
+    def _crawl(self) -> Iterable[Result]:
+        return []
```
[the remark this answers is not in the document]
Comment (ericvergnaud) [reply src/databricks/labs/ucx/source_code/directfs_access.py:123]: That works fine if the process is dedicated, but for jobs, the process collects simultaneously JobProblem's and DirectFsAccess's...

File: src/databricks/labs/ucx/source_code/directfs_access.py [hunk 18]
```diff
@@ -0,0 +1,136 @@
...
+        try:
+            self._append_records(dfsas)
+        except DatabricksError as e:
+            logger.error("Failed to store DFSAs", exc_info=e)
+
+    def _try_fetch(self) -> Iterable[DirectFsAccess]:
+        sql = f"SELECT * FROM {self.full_name}"
+        yield from self._backend.fetch(sql)
+
+    def _crawl(self) -> Iterable[Result]:
```
[the remark this answers is not in the document]
Comment (ericvergnaud) [reply src/databricks/labs/ucx/source_code/directfs_access.py:122]: Fixed

File: src/databricks/labs/ucx/source_code/directfs_access.py [hunk 19]
```diff
@@ -0,0 +1,136 @@
...
+            source_lineage=self.source_lineage,
+            job_id=self.job_id,
+            job_name=self.job_name,
+            task_key=self.task_key,
+            assessment_start_timestamp=assessment_start or self.assessment_start_timestamp,
+            assessment_end_timestamp=assessment_end or self.assessment_start_timestamp,
+        )
+
+
+class _DirectFsAccessCrawler(CrawlerBase):
```
[the remark this answers is not in the document]
Comment (ericvergnaud) [reply src/databricks/labs/ucx/source_code/directfs_access.py:100]: done

File: src/databricks/labs/ucx/source_code/directfs_access.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (ericvergnaud) [reply src/databricks/labs/ucx/source_code/directfs_access.py]: See above comment

File: src/databricks/labs/ucx/install.py [hunk 20]
```diff
@@ -120,6 +121,9 @@ def deploy_schema(sql_backend: SqlBackend, inventory_schema: str):
             functools.partial(table, "udfs", Udf),
             functools.partial(table, "logs", LogRecord),
             functools.partial(table, "recon_results", ReconResult),
+            functools.partial(
```
Comment (ericvergnaud) [inline src/databricks/labs/ucx/install.py:124]: Adding `direct_file_system_access_in_queries` table will be done in PR https://github.com/databrickslabs/ucx/pull/2599

File: src/databricks/labs/ucx/install.py [hunk 21]
```diff
@@ -128,6 +132,7 @@ def deploy_schema(sql_backend: SqlBackend, inventory_schema: str):
     deployer.deploy_view("misc_patterns", "queries/views/misc_patterns.sql")
     deployer.deploy_view("code_patterns", "queries/views/code_patterns.sql")
     deployer.deploy_view("reconciliation_results", "queries/views/reconciliation_results.sql")
+    # direct_file_system_access view will be added in upcoming PR
```
Comment (ericvergnaud) [inline src/databricks/labs/ucx/install.py:135]: Adding `direct_file_system_access` view will be done in PR https://github.com/databrickslabs/ucx/pull/2599

File: src/databricks/labs/ucx/source_code/jobs.py [hunk 22]
```diff
@@ -441,3 +493,85 @@ def _process_dependency(
         linter = FileLinter(ctx, path_lookup, self._session_state, dependency.path, inherited_tree)
         for advice in linter.lint():
             yield LocatedAdvice(advice, dependency.path)
+
+
+def _get_path_modified_datetime(path: Path) -> datetime:
+    unix_time = 0.0
+    if isinstance(path, WorkspacePath):
+        # TODO add stats method in blueprint, see https://github.com/databrickslabs/blueprint/issues/142
+        # pylint: disable=protected-access
```
Comment (ericvergnaud) [inline src/databricks/labs/ucx/source_code/jobs.py:502]: will go away with https://github.com/databrickslabs/blueprint/pull/144

File: src/databricks/labs/ucx/source_code/directfs_access.py [hunk 23]
```diff
@@ -0,0 +1,136 @@
...
+            sql_backend (SqlBackend): The SQL Execution Backend abstraction (either REST API or Spark)
+            schema: The schema name for the inventory persistence.
+        """
+        super().__init__(backend, "hive_metastore", schema, table, DirectFsAccess)
+
+    def append(self, dfsas: Sequence[DirectFsAccess]):
+        try:
+            self._append_records(dfsas)
+        except DatabricksError as e:
+            logger.error("Failed to store DFSAs", exc_info=e)
```
[the remark this answers is not in the document]
Comment (ericvergnaud) [reply src/databricks/labs/ucx/source_code/directfs_access.py]: We also need a crawler that is happy with existing empty snapshots

File: src/databricks/labs/ucx/install.py [hunk 24]
```diff
@@ -120,6 +121,9 @@ def deploy_schema(sql_backend: SqlBackend, inventory_schema: str):
             functools.partial(table, "udfs", Udf),
             functools.partial(table, "logs", LogRecord),
             functools.partial(table, "recon_results", ReconResult),
+            functools.partial(
+                table, "direct_file_system_access_in_paths", DirectFsAccess
```
Comment (nfx) [inline src/databricks/labs/ucx/install.py:125]:
```suggestion
                table, "directfs_in_workspace", DirectFsAccess
```
please pick a shorter name, as it's abnormally longer than others

File: src/databricks/labs/ucx/queries/views/direct_fs_access.sql [hunk 25]
```diff
@@ -0,0 +1,7 @@
+SELECT
+    *
```
Comment (nfx) [inline src/databricks/labs/ucx/queries/views/direct_fs_access.sql:2]: nit: as a good practice of views, please specify explicit column names

**Review by nfx — COMMENTED** [review]
waiting for final reply from @asnare

File: src/databricks/labs/ucx/queries/views/direct_fs_access.sql (hunk 25 above)
Comment (asnare) [reply src/databricks/labs/ucx/queries/views/direct_fs_access.sql]: … especially for `UNION` where the order of columns is important. 

(People often assume that schema changes which modify the column order are forward and backward compatible and don't realise `UNION`s will break. The only solution is to defensively enumerate columns explicitly when performing a `UNION`.)

File: src/databricks/labs/ucx/source_code/directfs_access.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (asnare) [reply src/databricks/labs/ucx/source_code/directfs_access.py:40]: I don't fully understand why `None` wouldn't be a better default value for these. Is the problem that we don't want to mark the column types as optional?

File: src/databricks/labs/ucx/queries/views/direct_fs_access.sql (hunk 25 above)
Comment (ericvergnaud) [reply src/databricks/labs/ucx/queries/views/direct_fs_access.sql:2]: dropped from this PR

File: src/databricks/labs/ucx/source_code/jobs.py [hunk 26]
```diff
@@ -441,3 +493,85 @@ def _process_dependency(
         linter = FileLinter(ctx, path_lookup, self._session_state, dependency.path, inherited_tree)
         for advice in linter.lint():
             yield LocatedAdvice(advice, dependency.path)
+
+
+def _get_path_modified_datetime(path: Path) -> datetime:
+    unix_time = 0.0
```
Comment (asnare) [inline src/databricks/labs/ucx/source_code/jobs.py:499]: All branches effectively just set a single value… I'm not sure why the current initialize/add branching control flow is in place?

File: src/databricks/labs/ucx/source_code/directfs_access.py [hunk 27]
```diff
@@ -0,0 +1,137 @@
...
+            schema: The schema name for the inventory persistence.
+        """
+        super().__init__(backend, "hive_metastore", schema, table, DirectFsAccess)
+
+    def dump_all(self, dfsas: Sequence[DirectFsAccess]):
+        try:
+            # TODO until we historize data, we append all DFSAs
+            self._update_snapshot(dfsas, mode="append")
+        except DatabricksError as e:
+            logger.error("Failed to store DFSAs", exc_info=e)
```
Comment (asnare) [inline src/databricks/labs/ucx/source_code/directfs_access.py:117]: I'd like to make it clear that this is only being called once per run with a batch of everything instead of incrementally with lots of small updates. Can we put a docstring or some form of internal documentation in place to explain that:

 * This crawler isn't using the normal pull-based pattern, but is rather push-based because multiple sets of data are being accumulated during a single (expensive) pass over the code being linted.
 * The records for a pass are written in a single batch. (New name is better, thanks. 👍)
 * Updating the crawler to support multiple sets of data is out of scope and will be the subject of a future PR.

File: src/databricks/labs/ucx/source_code/jobs.py [hunk 28]
```diff
@@ -69,6 +78,13 @@ def load(self, path_lookup: PathLookup) -> SourceContainer | None:
     def __repr__(self):
         return f'WorkflowTask<{self._task.task_key} of {self._job.settings.name}>'
 
+    @property
+    def lineage(self) -> list[LineageAtom]:
+        job_name = ("" if self._job.settings is None else self._job.settings.name) or "unknown job"
```
Comment (asnare) [inline src/databricks/labs/ucx/source_code/jobs.py:83]:
```suggestion
        job_name = (None if self._job.settings is None else self._job.settings.name) or "unknown job"
```

**Review by asnare — COMMENTED** [review]
Nothing major; mainly some clarifying remarks/questions.

File: src/databricks/labs/ucx/install.py (hunk 24 above)
Comment (ericvergnaud) [reply src/databricks/labs/ucx/install.py]: done, sticking to .`..paths` because `queries` also belong to workspace

File: src/databricks/labs/ucx/source_code/directfs_access.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (ericvergnaud) [reply src/databricks/labs/ucx/source_code/directfs_access.py:40]: Indeed, marking them as optional also affects the schema in a bad way, hence the proposed approach.

File: src/databricks/labs/ucx/source_code/directfs_access.py (hunk 27 above)
Comment (ericvergnaud) [reply src/databricks/labs/ucx/source_code/directfs_access.py:117]: Done

File: src/databricks/labs/ucx/source_code/jobs.py (hunk 26 above)
Comment (ericvergnaud) [reply src/databricks/labs/ucx/source_code/jobs.py]: Fixed

File: src/databricks/labs/ucx/source_code/directfs_access.py [hunk 29]
```diff
@@ -0,0 +1,141 @@
...
+    source_id: str = UNKNOWN
+    source_timestamp: datetime = datetime.fromtimestamp(0)
+    source_lineage: list[LineageAtom] = field(default_factory=list)
+    job_id: int = -1
+    job_name: str = UNKNOWN
+    task_key: str = UNKNOWN
+    assessment_start_timestamp: datetime = datetime.fromtimestamp(0)
+    assessment_end_timestamp: datetime = datetime.fromtimestamp(0)
+
+    def replace_source(
```
Comment (nfx) [inline src/databricks/labs/ucx/source_code/directfs_access.py:42]: in a follow-up PR, please use `dataclasses.replace`

**Review by nfx — APPROVED** [review]
lgtm
